File: phase4-hardware/arduino/Aeon.cpp
/**
 * Aeon.cpp - Implementación de librería Arduino
 *
 * (c) 2024 Sistemas Ursol - Jeremy Arias Solano
 */

#include "Aeon.h"

// Factor de escala para punto fijo Q8.8
#define SCALE 256
#define SCALE_BITS 8

Aeon::Aeon(uint8_t reservoirSize) {
  _size = min(reservoirSize, AEON_MAX_RESERVOIR);
  _trained = false;
  _sparse_count = 0;
  _rng = 0;
}

void Aeon::begin(uint32_t seed) {
  // Inicializar RNG
  _rng = (seed == 0) ? millis() : seed;

  // Limpiar estado
  reset();

  // Inicializar W_in (pesos de entrada aleatorios)
  for (uint8_t i = 0; i < _size; i++) {
    _W_in[i] = (int8_t)((_random() % 256) - 128);
    _W_out[i] = 0; // Se entrena después
  }

  // Inicializar reservoir escaso
  _sparse_count = 0;
  uint16_t target = (_size * _size) / AEON_SPARSITY;
  uint16_t max_sparse = sizeof(_sparse_from);

  for (uint16_t i = 0; i < target && _sparse_count < max_sparse; i++) {
    uint8_t from = _random() % _size;
    uint8_t to = _random() % _size;
    int8_t weight = (int8_t)((_random() % 256) - 128);

    _sparse_from[_sparse_count] = from;
    _sparse_to[_sparse_count] = to;
    _sparse_weight[_sparse_count] = weight;
    _sparse_count++;
  }

  _trained = false;
}

uint32_t Aeon::_random() {
  _rng = (_rng * 1103515245UL + 12345UL) & 0x7FFFFFFFUL;
  return _rng;
}

int16_t Aeon::_tanh_approx(int16_t x) {
  // Saturación y aproximación simple
  if (x > SCALE)
    return SCALE;
  if (x < -SCALE)
    return -SCALE;

  // tanh(x) ≈ x - x³/3 para |x| < 1
  int32_t x2 = ((int32_t)x * x) >> SCALE_BITS;
  int32_t x3 = (x2 * x) >> SCALE_BITS;
  return x - (x3 / 3);
}

float Aeon::_toFloat(int16_t fixed) { return (float)fixed / SCALE; }

int16_t Aeon::_toFixed(float f) { return (int16_t)(f * SCALE); }

void Aeon::update(float input) {
  int16_t input_fixed = _toFixed(input);
  int16_t new_state[AEON_MAX_RESERVOIR];

  // Contribución de entrada
  for (uint8_t i = 0; i < _size; i++) {
    int32_t sum = ((int32_t)_W_in[i] * input_fixed) >> SCALE_BITS;
    new_state[i] = (int16_t)sum;
  }

  // Contribución del reservoir (escaso)
  for (uint16_t k = 0; k < _sparse_count; k++) {
    uint8_t from = _sparse_from[k];
    uint8_t to = _sparse_to[k];
    int32_t contrib = ((int32_t)_sparse_weight[k] * _state[from]) >> SCALE_BITS;
    new_state[to] += (int16_t)contrib;
  }

  // Aplicar tanh y actualizar
  for (uint8_t i = 0; i < _size; i++) {
    _state[i] = _tanh_approx(new_state[i]);
  }
}

float Aeon::predict() {
  int32_t sum = 0;
  for (uint8_t i = 0; i < _size; i++) {
    sum += ((int32_t)_W_out[i] * _state[i]) >> SCALE_BITS;
  }
  return _toFloat((int16_t)sum);
}

float Aeon::train(float *inputs, float *targets, uint16_t n_samples,
                  uint8_t washout) {
  if (n_samples <= washout)
    return -1.0f;

  uint16_t train_samples = n_samples - washout;

  // Acumuladores para regresión simplificada
  float sum_xy[AEON_MAX_RESERVOIR];
  float sum_xx[AEON_MAX_RESERVOIR];

  for (uint8_t i = 0; i < _size; i++) {
    sum_xy[i] = 0;
    sum_xx[i] = 0;
  }

  // Reset y recolectar
  reset();

  for (uint16_t t = 0; t < n_samples; t++) {
    update(inputs[t]);

    if (t >= washout) {
      float y = targets[t];
      for (uint8_t i = 0; i < _size; i++) {
        float x = _toFloat(_state[i]);
        sum_xy[i] += x * y;
        sum_xx[i] += x * x;
      }
    }
  }

  // Calcular pesos de salida
  for (uint8_t i = 0; i < _size; i++) {
    float weight = sum_xy[i] / (sum_xx[i] + 1e-6f);
    weight = constrain(weight, -2.0f, 2.0f);
    _W_out[i] = (int8_t)(weight * 64); // Escala reducida para int8
  }

  _trained = true;

  // Calcular MSE
  reset();
  float mse = 0;

  for (uint16_t t = washout; t < n_samples; t++) {
    update(inputs[t]);
    float pred = predict();
    float diff = pred - targets[t];
    mse += diff * diff;
  }

  return mse / train_samples;
}

void Aeon::reset() {
  for (uint8_t i = 0; i < _size; i++) {
    _state[i] = 0;
  }
}

uint16_t Aeon::memoryUsage() {
  return sizeof(Aeon) + _size * sizeof(int16_t) + // state
         _size * sizeof(int8_t) * 2 +             // W_in, W_out
         _sparse_count * 3;                       // sparse connections
}

File: phase3-integration/aeon.js
/**
 * Proyecto Eón - AeonJS
 * Implementación JavaScript pura del núcleo ESN
 *
 * "La inteligencia no se crea, se descubre."
 *
 * Esta versión es compatible con navegadores sin necesidad de WASM.
 * Para producción, usar la versión compilada con Emscripten.
 */

class AeonCore {
  /**
   * Configuración por defecto
   */
  static DEFAULT_CONFIG = {
    reservoirSize: 32,
    inputSize: 1,
    outputSize: 1,
    sparsityFactor: 4,
    spectralRadius: 0.9,
  };

  /**
   * Crea una nueva instancia de Eón (Momento Cero)
   */
  constructor(config = {}) {
    this.config = { ...AeonCore.DEFAULT_CONFIG, ...config };

    // === MOMENTO CERO ===
    this.birthTimestamp = new Date();
    this.birthHash = this._generateHash();
    this.reservoirSeed = Date.now() % 0xffffffff;

    // Inicializar RNG
    this._rngState = this.reservoirSeed;

    // Estado del reservoir
    this.state = new Float32Array(this.config.reservoirSize);

    // Matrices de pesos
    this.W_in = new Float32Array(
      this.config.reservoirSize * this.config.inputSize
    );
    this.W_out = new Float32Array(
      this.config.outputSize * this.config.reservoirSize
    );

    // Conexiones escasas del reservoir
    this.sparseIndices = [];
    this.sparseWeights = [];

    // Estado de aprendizaje
    this.isTrained = false;
    this.samplesProcessed = 0;
    this.learningSessions = 0;

    // Inicializar reservoir
    this._initializeReservoir();
  }

  /**
   * Genera hash único de nacimiento
   */
  _generateHash() {
    const data = `${Date.now()}-${Math.random()}-${
      navigator?.userAgent || "node"
    }`;
    let hash = 0;
    for (let i = 0; i < data.length; i++) {
      const char = data.charCodeAt(i);
      hash = (hash << 5) - hash + char;
      hash = hash & hash;
    }
    return Math.abs(hash).toString(16).padStart(16, "0").slice(0, 16);
  }

  /**
   * Generador LCG
   */
  _random() {
    this._rngState = (this._rngState * 1103515245 + 12345) & 0x7fffffff;
    return this._rngState / 0x7fffffff;
  }

  /**
   * Inicializa el reservoir aleatorio
   */
  _initializeReservoir() {
    const { reservoirSize, inputSize, sparsityFactor } = this.config;

    // W_in: Pesos de entrada aleatorios [-1, 1]
    for (let i = 0; i < reservoirSize * inputSize; i++) {
      this.W_in[i] = this._random() * 2 - 1;
    }

    // W_reservoir: Conexiones escasas
    const totalConnections = reservoirSize * reservoirSize;
    const targetConnections = Math.floor(totalConnections / sparsityFactor);

    const usedIndices = new Set();
    for (let i = 0; i < targetConnections; i++) {
      let idx;
      do {
        idx = Math.floor(this._random() * totalConnections);
      } while (usedIndices.has(idx));

      usedIndices.add(idx);
      this.sparseIndices.push(idx);
      this.sparseWeights.push(this._random() * 2 - 1);
    }

    // Escalar pesos para radio espectral (aproximación)
    const scale =
      this.config.spectralRadius / Math.sqrt(targetConnections / reservoirSize);
    for (let i = 0; i < this.sparseWeights.length; i++) {
      this.sparseWeights[i] *= scale;
    }
  }

  /**
   * Aproximación rápida de tanh
   */
  _tanh(x) {
    if (x > 2) return 1;
    if (x < -2) return -1;
    const x2 = x * x;
    return x * (1 - x2 / 3 + (x2 * x2) / 15);
  }

  /**
   * Actualiza el estado del reservoir
   */
  update(input) {
    const { reservoirSize, inputSize } = this.config;
    const newState = new Float32Array(reservoirSize);

    // Contribución de entrada
    for (let i = 0; i < reservoirSize; i++) {
      let sum = 0;
      for (let j = 0; j < inputSize; j++) {
        sum += this.W_in[i * inputSize + j] * input[j];
      }
      newState[i] = sum;
    }

    // Contribución del reservoir (escaso)
    for (let k = 0; k < this.sparseIndices.length; k++) {
      const idx = this.sparseIndices[k];
      const i = Math.floor(idx / reservoirSize);
      const j = idx % reservoirSize;
      newState[i] += this.sparseWeights[k] * this.state[j];
    }

    // Aplicar no-linealidad
    for (let i = 0; i < reservoirSize; i++) {
      this.state[i] = this._tanh(newState[i]);
    }

    this.samplesProcessed++;
  }

  /**
   * Genera predicción
   */
  predict() {
    const { reservoirSize, outputSize } = this.config;
    const output = new Float32Array(outputSize);

    for (let o = 0; o < outputSize; o++) {
      let sum = 0;
      for (let j = 0; j < reservoirSize; j++) {
        sum += this.W_out[o * reservoirSize + j] * this.state[j];
      }
      output[o] = sum;
    }

    return output;
  }

  /**
   * Entrena con datos
   */
  train(inputs, targets, washout = 50) {
    if (inputs.length <= washout) {
      throw new Error("Insufficient training data");
    }

    const { reservoirSize, outputSize } = this.config;
    const trainSamples = inputs.length - washout;

    // Reset
    this.reset();

    // Acumuladores para regresión
    const StS = [];
    const StY = [];

    for (let i = 0; i < reservoirSize; i++) {
      StS[i] = new Float32Array(reservoirSize);
      StS[i][i] = 1e-4; // Regularización
      StY[i] = new Float32Array(outputSize);
    }

    // Recolectar estados
    for (let t = 0; t < inputs.length; t++) {
      this.update([inputs[t]]);

      if (t >= washout) {
        const target = targets[t];

        // Acumular S^T * S y S^T * Y
        for (let i = 0; i < reservoirSize; i++) {
          for (let j = i; j < reservoirSize; j++) {
            const prod = this.state[i] * this.state[j];
            StS[i][j] += prod;
            if (i !== j) StS[j][i] += prod;
          }
          for (let o = 0; o < outputSize; o++) {
            StY[i][o] += this.state[i] * target;
          }
        }
      }
    }

    // Invertir StS usando Gauss-Jordan
    const inv = [];
    for (let i = 0; i < reservoirSize; i++) {
      inv[i] = new Float32Array(reservoirSize);
      inv[i][i] = 1;
    }

    for (let col = 0; col < reservoirSize; col++) {
      // Pivoteo
      let maxRow = col;
      for (let row = col + 1; row < reservoirSize; row++) {
        if (Math.abs(StS[row][col]) > Math.abs(StS[maxRow][col])) {
          maxRow = row;
        }
      }

      if (maxRow !== col) {
        [StS[col], StS[maxRow]] = [StS[maxRow], StS[col]];
        [inv[col], inv[maxRow]] = [inv[maxRow], inv[col]];
      }

      const pivot = StS[col][col] || 1e-10;
      for (let k = 0; k < reservoirSize; k++) {
        StS[col][k] /= pivot;
        inv[col][k] /= pivot;
      }

      for (let row = 0; row < reservoirSize; row++) {
        if (row !== col) {
          const factor = StS[row][col];
          for (let k = 0; k < reservoirSize; k++) {
            StS[row][k] -= factor * StS[col][k];
            inv[row][k] -= factor * inv[col][k];
          }
        }
      }
    }

    // Calcular W_out = inv * StY
    for (let o = 0; o < outputSize; o++) {
      for (let i = 0; i < reservoirSize; i++) {
        let sum = 0;
        for (let k = 0; k < reservoirSize; k++) {
          sum += inv[i][k] * StY[k][o];
        }
        // Limitar magnitud
        sum = Math.max(-2, Math.min(2, sum));
        this.W_out[o * reservoirSize + i] = sum;
      }
    }

    this.isTrained = true;
    this.learningSessions++;

    // Calcular MSE
    this.reset();
    let mse = 0;

    for (let t = washout; t < inputs.length; t++) {
      this.update([inputs[t]]);
      const pred = this.predict();
      const diff = pred[0] - targets[t];
      mse += diff * diff;
    }

    return mse / trainSamples;
  }

  /**
   * Resetea estado
   */
  reset() {
    this.state.fill(0);
  }

  /**
   * Edad en segundos
   */
  get ageSeconds() {
    return Math.floor((Date.now() - this.birthTimestamp.getTime()) / 1000);
  }

  /**
   * Memoria aproximada en bytes
   */
  get memoryBytes() {
    return (
      this.W_in.byteLength +
      this.W_out.byteLength +
      this.state.byteLength +
      this.sparseWeights.length * 4 +
      this.sparseIndices.length * 4
    );
  }

  /**
   * Estado serializable
   */
  getStatus() {
    return {
      name: `Aeon-${this.birthHash.slice(0, 8)}`,
      birthTimestamp: this.birthTimestamp.toISOString(),
      birthHash: this.birthHash,
      age: this._formatAge(),
      ageSeconds: this.ageSeconds,
      isTrained: this.isTrained,
      samplesProcessed: this.samplesProcessed,
      learningSessions: this.learningSessions,
      reservoirSize: this.config.reservoirSize,
      memoryKB: (this.memoryBytes / 1024).toFixed(2),
      sparseConnections: this.sparseIndices.length,
    };
  }

  _formatAge() {
    const s = this.ageSeconds;
    if (s < 60) return `${s} segundos`;
    if (s < 3600) return `${(s / 60).toFixed(1)} minutos`;
    if (s < 86400) return `${(s / 3600).toFixed(1)} horas`;
    return `${(s / 86400).toFixed(1)} días`;
  }

  /**
   * Exportar a JSON
   */
  toJSON() {
    return {
      config: this.config,
      birthTimestamp: this.birthTimestamp.toISOString(),
      birthHash: this.birthHash,
      reservoirSeed: this.reservoirSeed,
      W_in: Array.from(this.W_in),
      W_out: Array.from(this.W_out),
      sparseIndices: this.sparseIndices,
      sparseWeights: this.sparseWeights,
      isTrained: this.isTrained,
      samplesProcessed: this.samplesProcessed,
      learningSessions: this.learningSessions,
    };
  }

  /**
   * Cargar desde JSON
   */
  static fromJSON(data) {
    const core = new AeonCore(data.config);
    core.birthTimestamp = new Date(data.birthTimestamp);
    core.birthHash = data.birthHash;
    core.reservoirSeed = data.reservoirSeed;
    core.W_in = new Float32Array(data.W_in);
    core.W_out = new Float32Array(data.W_out);
    core.sparseIndices = data.sparseIndices;
    core.sparseWeights = data.sparseWeights;
    core.isTrained = data.isTrained;
    core.samplesProcessed = data.samplesProcessed;
    core.learningSessions = data.learningSessions;
    return core;
  }
}

// Exportar para Node.js y navegador
if (typeof module !== "undefined" && module.exports) {
  module.exports = { AeonCore };
}
if (typeof window !== "undefined") {
  window.AeonCore = AeonCore;
}
